RecoEgamma/ElectronIdentification/plugins/ElectronMVAVariableHelper.cc

```diff
@@ -1,10 +1,119 @@
+{
+  produces<edm::ValueMap<float>>("convVtxFitProb");
+  produces<edm::ValueMap<float>>("kfhits");
+  produces<edm::ValueMap<float>>("kfchi2");
+}
+
+ElectronMVAVariableHelper::~ElectronMVAVariableHelper()
+{}
+
+void ElectronMVAVariableHelper::produce(edm::Event & iEvent, const edm::EventSetup & iSetup) {
+
+  // get Handles
+  auto electrons      = electronsToken_  .getValidHandle<edm::View<reco::GsfElectron>>(iEvent);
+  auto vtxH           = vtxToken_        .getValidHandle<reco::VertexCollection>      (iEvent);
+  auto conversions    = conversionsToken_.getValidHandle<reco::ConversionCollection>  (iEvent);
+  auto beamSpotHandle = beamSpotToken_   .getValidHandle<reco::BeamSpot>              (iEvent);
+
+  const reco::VertexRef vtx(vtxH, 0);
+  const reco::BeamSpot* beamSpot = &*(beamSpotHandle.product());
+
+  // prepare vector for output
+  std::vector<float> convVtxFitProbVals;
+  std::vector<float> kfhitsVals;
+  std::vector<float> kfchi2Vals;
+
+  for (size_t i = 0; i < electrons->size(); ++i){
+      auto iCand = electrons->ptrAt(i);
+
+      // Conversion vertex fit
+      reco::ConversionRef convRef = ConversionTools::matchedConversion(*iCand, conversions, beamSpot->position());
+
+      float convVtxFitProb = -1.;
+      if(!convRef.isNull()) {
+          const reco::Vertex &vtx = convRef.get()->conversionVertex();
+          if (vtx.isValid()) {
+              convVtxFitProb = TMath::Prob( vtx.chi2(),  vtx.ndof());
+          }
+      }
+
+      convVtxFitProbVals.push_back(convVtxFitProb);
+
+      // kf track related variables
+      bool validKf=false;
+      reco::TrackRef trackRef = iCand->closestCtfTrackRef();
+      validKf = trackRef.isAvailable();
+      validKf &= trackRef.isNonnull();
+      float kfchi2 = validKf ? trackRef->normalizedChi2() : 0 ; //ielectron->track()->normalizedChi2() : 0 ;
+      float kfhits = validKf ? trackRef->hitPattern().trackerLayersWithMeasurement() : -1. ;
+
+      kfchi2Vals.push_back(kfchi2);
+      kfhitsVals.push_back(kfhits);
+
+  }
+
+  // convert into ValueMap and store
+  writeValueMap(iEvent, electrons, kfchi2Vals, "kfchi2" );
+  writeValueMap(iEvent, electrons, kfhitsVals, "kfhits" );
+  writeValueMap(iEvent, electrons, convVtxFitProbVals, "convVtxFitProb" );
+}
 
-typedef ElectronMVAVariableHelper<reco::GsfElectron> GsfElectronMVAVariableHelper;
+void ElectronMVAVariableHelper::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  // electronMVAVariableHelper
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("conversions",             edm::InputTag("allConversions"));
+  desc.add<edm::InputTag>("src",                     edm::InputTag("gedGsfElectrons"));
+  desc.add<edm::InputTag>("srcMiniAOD",              edm::InputTag("slimmedElectrons","","@skipCurrentProcess"));
+  desc.add<edm::InputTag>("beamSpot",                edm::InputTag("offlineBeamSpot"));
+  desc.add<edm::InputTag>("beamSpotMiniAOD",         edm::InputTag("offlineBeamSpot"));
```

Comment:
it looks like there was no actual reason to make beamSpot and beamSpotMiniAOD. Do we ever have an alternative ```offlineBeamSpot``` ?